src/pages.c

```diff
@@ -567,6 +567,27 @@ pages_nohuge_unaligned(void *addr, size_t size) {
 	return pages_nohuge_impl(addr, size, false);
 }
 
+bool
+pages_collapse(void *addr, size_t size) {
+	assert(PAGE_ADDR2BASE(addr) == addr);
+	assert(PAGE_CEILING(size) == size);
+	/*
+	 * There is one more MADV_COLLAPSE precondition that is not easy to
+	 * express with assert statement. In order to madvise(addr, size,
+	 * MADV_COLLAPSE) call to be successful, at least one page in the range
+	 * must currently be backed by physical memory.
+	 */
+#if defined(JEMALLOC_HAVE_MADVISE_COLLAPSE)
+#  ifdef MADV_COLLAPSE
+	return (madvise(addr, size, MADV_COLLAPSE) != 0);
+#  else
+	return (madvise(addr, size, JEMALLOC_MADV_COLLAPSE) != 0);
```

Comment:
Does this means if `MADV_COLLAPSE` is not defined, i.e., in Linux kernel before 6.1, we can still use the arbitrary number 25 in the madvise call to get exactly the same outcome as `MADV_COLLAPSE` on x86 machines?

Reply:
> Does this means if MADV_COLLAPSE is not defined, i.e., in Linux kernel before 6.1, we can still use the arbitrary number 25 in the madvise call to get exactly the same outcome as MADV_COLLAPSE on x86 machines?

There are are few different cases as we have couple of moving parts to consider: Linux kernel might or might not support `MADV_COLLAPSE` feature and we might have libc with or without `MADV_COLLAPSE` constant defined ([patch][1] for glibc defined constant).

1. If Linux kernel is older than 6.1, then we'll get `EINVAL` from `madvise(addr, size, MADV_COLLAPSE)` call.
2. If Linux kernel in newer than 6.1, but libc doesn't have `MADV_COLLAPSE` constant defined. `madvise(addr, size, MADV_COLLAPSE)` will not compile, but `madvise(addr, size, JEMALLOC_MADV_COLLAPSE /* 25 */)` will and call will be successful (subject to hugepages availability in the system).
3.  Linux kernel in newer than 6.1 and `MADV_COLLAPSE` constant defined.

Internally, we actually have case 2.

There is also an alternative solution: do not define `JEMALLOC_DEFINE_MADVISE_COLLAPSE` in autoconf at all, but allow users to define this macros at compile time manually, if they know target system supports `MADV_COLLAPSE` feature, but libc doesn't have constant defined. Actually, I think it is a better solution to be honest, but I implemented it the way it is implemented now to match similar logic for `MADV_FREE`.

https://github.com/jemalloc/jemalloc/blob/2a693b83d2d1631b6a856d178125e1c47c12add9/configure.ac#L2427-L2444

Counter-argument is feature is guarded by `hpa_hugify_sync` flag, which is off by default and we probably don't need additional guarding on compile time. Worst case scenario, user will enable `hpa_hugify_sync` option on unsupported system and we'll fallback to synchronous hugification in 100% of cases always incrementing `nhugify_failures` counter.

[1]: https://sourceware.org/pipermail/libc-alpha/2022-December/144241.html

Reply:
> jemalloc cannot easily know whether "at least one page in the range is backed by physical memory".

"at least one page in the range is backed by physical memory" is a quote from `MADV_COLLAPSE`
 [documentation][1]. Full quote is:

> All non-resident pages covered by the range will first be swapped/faulted-in, before being copied onto a freshly allocated hugepage.  If the native pages compose the same PTE-mapped hugepage, and are suitably aligned, allocation of a new hugepage may be elided and collapse may happen in-place.  Unmapped pages will have their data directly initialized to 0 in the new hugepage.  However, for every eligible hugepage-aligned/sized region to be collapsed, at least one page must currently be backed by physical memory.

Based on full text I think «backed by physical memory» piece means: at least one page in the range should be faulted in. Given, right now we only hugify pages with more than `hpa_hugification_threshold` bytes allocated, I am not sure this part is relevant for us right now, but I still think comment might be useful in case we'll change hugification policy in the future. Probably, we should change wording a bit and use «at least one page was faulted in» instead of «backed by physical memory», because without the full context someone might read it a bit differently.

>  If this condition is not satisfied, the call to pages_collapse would fail and fall back to the async hugification. My question is: how much is the cost for the fallback

Fallback call of `madvise(addr, size, MADV_HUGEPAGE)` flips kernel flag, so synchronous part of it should be quite cheap. There are some asynchronous work done, but we are doing it right now anyway. One relevant concern is: if 
`madvise(addr, size, MADV_COLLAPSE)` failed, because there is no hugepages available, I think there is a high probability `madvise(addr, size, MADV_HUGEPAGE)` will fail as well, but I guess we still might hugify page eventually if something changes in the system.

> and how much win we can expect if the synchronous one always succeed?

We observed two digit percent differences in hugepages coverage internally for 3 different services I tried to use `MADV_COLLAPSE` instead of `MADV_HUGEPAGE`.

> Following that, is there a way to heuristically know that MADV_COLLAPSE would succeed in a relatively high ratio, e.g., small sizes? I am thinking maybe instead of replacing all async hugificaiton to the sync one when the option is enabled, replacing it only on the occasions where we are more confident might bring about larger benefits.

I think `MADV_COLLAPSE` success rate is more of a function of physical memory fragmentation on the kernel level, that something we can observe within userspace memory allocator. In other words, if system is in a good shape, `MADV_COLLAPSE` will likely succeed, but if physical memory is highly fragmented and kernel can't do anything about it, then `MADV_COLLAPSE` call will fail.

[1]: https://man7.org/linux/man-pages/man2/madvise.2.html

Reply:
Now I am having doubts about fallback behaviour. There are different things can go wrong: some of them might be resolved with retry, but seems there are cases, when even khugepaged couldn't help much.

https://github.com/torvalds/linux/blob/2e1b3cc9d7f790145a80cb705b168f05dab65df2/mm/khugepaged.c#L2676-L2704

I'll think about it a bit more.